pkg/actions/mustgather.go

```diff
@@ -0,0 +1,364 @@
+				header.Name = strings.Replace(path, mustGatherDirName+string(os.PathSeparator), "", 1)
+
+				// Change to deflate to gain better compression
+				// see http://golang.org/pkg/archive/zip/#pkg-constants
+				header.Method = zip.Deflate
+
+				writer, err := zipWriter.CreateHeader(header)
+				if err != nil {
+					return err
+				}
+				_, err = io.Copy(writer, fileToZip)
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			logMG("walk error " + err.Error())
+		}
+	}
+}
+
+//getContainerID - returns the ID of the container filtered by name
+func getContainerID(containerName string) string {
+	dockerClient, dockerErr := docker.NewDockerClient()
+	if dockerErr != nil {
+		HandleDockerError(dockerErr)
+		os.Exit(1)
+	}
+	nameFilter := filters.NewArgs(filters.Arg("name", containerName))
+	container, getErr := docker.GetContainerListWithOptions(dockerClient, types.ContainerListOptions{All: true, Filters: nameFilter})
+	if getErr != nil {
+		HandleDockerError(getErr)
+		os.Exit(1)
+	}
+	return container[0].ID
+}
+
+//writeContainerInspectToFile - writes the results of `docker inspect containerId` to a file
+func writeContainerInspectToFile(containerID, containerName string) error {
+	dockerClient, dockerErr := docker.NewDockerClient()
+	if dockerErr != nil {
+		HandleDockerError(dockerErr)
+		os.Exit(1)
+	}
+	inspectedContents, inspectErr := docker.InspectContainer(dockerClient, containerID)
+	if inspectErr != nil {
+		HandleDockerError(inspectErr)
+		os.Exit(1)
+	}
+	fileContents, _ := json.MarshalIndent(inspectedContents, "", " ")
+	err := ioutil.WriteFile(filepath.Join(mustGatherDirName, containerName+".inspect"), fileContents, 0644)
+	return err
+}
+
+//writeContainerLogToFile - writes the results of `docker logs containerId` to a file
+func writeContainerLogToFile(containerID, containerName string) error {
+	dockerClient, dockerErr := docker.NewDockerClient()
+	if dockerErr != nil {
+		HandleDockerError(dockerErr)
+		os.Exit(1)
+	}
+	logStream, logErr := docker.GetContainerLogs(dockerClient, containerID)
+	if logErr != nil {
+		HandleDockerError(logErr)
+		os.Exit(1)
+	}
+	outFile, createErr := os.Create(filepath.Join(mustGatherDirName, containerName+".log"))
+	if createErr != nil {
+		errors.CheckErr(createErr, 201, "")
+	}
+	defer outFile.Close()
+	_, err := io.Copy(outFile, logStream)
+	return err
+}
+
+//copyCodewindWorkspace - copies the Codewind PFE container's workspace to mustgather
+func copyCodewindWorkspace(containerID string) error {
+	dockerClient, dockerErr := docker.NewDockerClient()
+	if dockerErr != nil {
+		HandleDockerError(dockerErr)
+		os.Exit(1)
+	}
+	tarFileStream, fileErr := docker.GetFilesFromContainer(dockerClient, containerID, "/codewind-workspace")
+	if fileErr != nil {
+		HandleDockerError(fileErr)
+		os.Exit(1)
+	}
+	defer tarFileStream.Close()
+	// Extracting tarred files
+	tarBallReader := tar.NewReader(tarFileStream)
```

Comment:
Could the tar reading part of this be merged with the code we have in filesystem.go UnTar
https://github.com/eclipse/codewind-installer/blob/18d4b2bd1087017b0957ec07932f38e66db578cc/pkg/utils/filesystem.go#L151
then we only have to have the code to handle properly reading a tar file in one place.

Reply:
Agreed